File: http/source/crypto_identity_capsule.c
#include "crypto_identity_capsule.h"
#include "crypto_openssl.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int is_forbidden_sensitive_name(const char *name)
{
    static const char *const forbidden[] = {
        "national_id", "national_id_number", "phone_number",
        "telephone_number", "dna_id", "biometric_id", "raw_dna", NULL
    };
    size_t i;

    if (name == NULL) {
        return 1;
    }
    for (i = 0U; forbidden[i] != NULL; ++i) {
        if (strcmp(name, forbidden[i]) == 0) {
            return 1;
        }
    }
    return 0;
}

int http3_cic_validate_metadata(const http3_cic_metadata_t *metadata,
                                uint64_t current_time)
{
    if (metadata == NULL || metadata->expiration <= current_time ||
        metadata->jurisdiction_id[0] == '\0' ||
        metadata->sender_reference[0] == '\0' ||
        metadata->emergency_endpoint_reference[0] == '\0') {
        return -1;
    }
    return 0;
}

int http3_cic_create_id(uint8_t capsule_id[HTTP3_CIC_ID_SIZE])
{
    return http3_openssl_random(capsule_id, HTTP3_CIC_ID_SIZE);
}

int http3_cic_reject_sensitive_attribute_name(const char *name)
{
    return is_forbidden_sensitive_name(name) ? -1 : 0;
}

int http3_cic_bind_context(const uint8_t capsule_id[HTTP3_CIC_ID_SIZE],
                          const char *jurisdiction_id,
                          const uint8_t *context,
                          size_t context_len,
                          uint8_t *output,
                          size_t output_capacity,
                          size_t *output_len)
{
    size_t jurisdiction_len;

    if (capsule_id == NULL || jurisdiction_id == NULL ||
        (context == NULL && context_len != 0U) || output == NULL ||
        output_len == NULL) {
        return -1;
    }

    jurisdiction_len = strlen(jurisdiction_id);
    if (jurisdiction_len >= HTTP3_CIC_MAX_JURISDICTION ||
        HTTP3_CIC_ID_SIZE + jurisdiction_len + context_len > output_capacity) {
        return -1;
    }

    memcpy(output, capsule_id, HTTP3_CIC_ID_SIZE);
    memcpy(output + HTTP3_CIC_ID_SIZE, jurisdiction_id, jurisdiction_len);
    if (context_len != 0U) {
        memcpy(output + HTTP3_CIC_ID_SIZE + jurisdiction_len, context, context_len);
    }
    *output_len = HTTP3_CIC_ID_SIZE + jurisdiction_len + context_len;
    return 0;
}

static int make_aad(const http3_cic_capsule_t *capsule,
                    uint8_t aad[HTTP3_CIC_ID_SIZE + HTTP3_CIC_MAX_JURISDICTION],
                    size_t *aad_len)
{
    size_t jurisdiction_len;

    if (capsule == NULL || aad == NULL || aad_len == NULL) {
        return -1;
    }
    jurisdiction_len = strlen(capsule->jurisdiction_id);
    if (jurisdiction_len >= HTTP3_CIC_MAX_JURISDICTION) {
        return -1;
    }
    memcpy(aad, capsule->capsule_id, HTTP3_CIC_ID_SIZE);
    memcpy(aad + HTTP3_CIC_ID_SIZE, capsule->jurisdiction_id, jurisdiction_len);
    *aad_len = HTTP3_CIC_ID_SIZE + jurisdiction_len;
    return 0;
}

int http3_cic_encrypt(const http3_cic_metadata_t *metadata,
                      const uint8_t recipient_public_key[32],
                      const uint8_t *plaintext,
                      size_t plaintext_len,
                      http3_cic_capsule_t *capsule)
{
    uint8_t ephemeral_private[32];
    uint8_t shared_secret[32];
    uint8_t key[32];
    uint8_t aad[HTTP3_CIC_ID_SIZE + HTTP3_CIC_MAX_JURISDICTION];
    size_t aad_len = 0U;
    size_t total_len;
    int rc = -1;

    memset(ephemeral_private, 0, sizeof(ephemeral_private));
    memset(shared_secret, 0, sizeof(shared_secret));
    memset(key, 0, sizeof(key));

    if (metadata == NULL || recipient_public_key == NULL || plaintext == NULL ||
        capsule == NULL || plaintext_len > SIZE_MAX - HTTP3_CIC_TAG_SIZE ||
        http3_cic_validate_metadata(metadata, 0U) != 0) {
        return -1;
    }

    memset(capsule, 0, sizeof(*capsule));
    capsule->version = HTTP3_CIC_VERSION;
    memcpy(capsule->capsule_id, metadata->capsule_id, HTTP3_CIC_ID_SIZE);
    memcpy(capsule->jurisdiction_id, metadata->jurisdiction_id,
           sizeof(capsule->jurisdiction_id));

    if (http3_openssl_x25519_generate(ephemeral_private,
                                      capsule->ephemeral_public_key) != 0 ||
        http3_openssl_x25519_derive(ephemeral_private, recipient_public_key,
                                    shared_secret) != 0 ||
        http3_openssl_hkdf_sha256(shared_secret, sizeof(shared_secret),
                                  metadata->capsule_id, HTTP3_CIC_ID_SIZE,
                                  (const uint8_t *)"SLeeLa/HTTP3/CIC/v1",
                                  sizeof("SLeeLa/HTTP3/CIC/v1") - 1U, key) != 0 ||
        http3_openssl_random(capsule->nonce, sizeof(capsule->nonce)) != 0 ||
        make_aad(capsule, aad, &aad_len) != 0) {
        goto done;
    }

    total_len = plaintext_len + HTTP3_CIC_TAG_SIZE;
    capsule->ciphertext = (uint8_t *)malloc(total_len);
    if (capsule->ciphertext == NULL) {
        goto done;
    }

    if (http3_openssl_aes256gcm_encrypt(
            key, capsule->nonce, aad, aad_len, plaintext, plaintext_len,
            capsule->ciphertext, capsule->ciphertext + plaintext_len) != 0) {
        http3_cic_free_capsule(capsule);
        goto done;
    }

    capsule->ciphertext_len = total_len;
    rc = 0;

done:
    memset(ephemeral_private, 0, sizeof(ephemeral_private));
    memset(shared_secret, 0, sizeof(shared_secret));
    memset(key, 0, sizeof(key));
    return rc;
}

int http3_cic_decrypt(const http3_cic_capsule_t *capsule,
                      const uint8_t recipient_private_key[32],
                      uint8_t *plaintext,
                      size_t plaintext_capacity,
                      size_t *plaintext_len)
{
    uint8_t shared_secret[32];
    uint8_t key[32];
    uint8_t aad[HTTP3_CIC_ID_SIZE + HTTP3_CIC_MAX_JURISDICTION];
    size_t aad_len = 0U;
    size_t data_len;
    int rc = -1;

    memset(shared_secret, 0, sizeof(shared_secret));
    memset(key, 0, sizeof(key));

    if (capsule == NULL || recipient_private_key == NULL || plaintext == NULL ||
        plaintext_len == NULL || capsule->version != HTTP3_CIC_VERSION ||
        capsule->ciphertext == NULL || capsule->ciphertext_len < HTTP3_CIC_TAG_SIZE ||
        capsule->ciphertext_len - HTTP3_CIC_TAG_SIZE > plaintext_capacity ||
        make_aad(capsule, aad, &aad_len) != 0) {
        return -1;
    }

    data_len = capsule->ciphertext_len - HTTP3_CIC_TAG_SIZE;
    if (http3_openssl_x25519_derive(recipient_private_key,
                                    capsule->ephemeral_public_key,
                                    shared_secret) != 0 ||
        http3_openssl_hkdf_sha256(shared_secret, sizeof(shared_secret),
                                  capsule->capsule_id, HTTP3_CIC_ID_SIZE,
                                  (const uint8_t *)"SLeeLa/HTTP3/CIC/v1",
                                  sizeof("SLeeLa/HTTP3/CIC/v1") - 1U, key) != 0 ||
        http3_openssl_aes256gcm_decrypt(
            key, capsule->nonce, aad, aad_len, capsule->ciphertext, data_len,
            capsule->ciphertext + data_len, plaintext) != 0) {
        memset(plaintext, 0, plaintext_capacity);
        goto done;
    }

    *plaintext_len = data_len;
    rc = 0;

done:
    memset(shared_secret, 0, sizeof(shared_secret));
    memset(key, 0, sizeof(key));
    return rc;
}

void http3_cic_free_capsule(http3_cic_capsule_t *capsule)
{
    if (capsule == NULL) {
        return;
    }
    if (capsule->ciphertext != NULL) {
        memset(capsule->ciphertext, 0, capsule->ciphertext_len);
        free(capsule->ciphertext);
    }
    memset(capsule, 0, sizeof(*capsule));
}

File: http/source/crypto_identity_capsule.py
"""Cryptographic Identity Capsule (CIC) reference implementation.

Dependencies:
    pip install cryptography

The implementation deliberately stores references to sensitive identity
records instead of raw National ID, telephone, or biometric/DNA data.

This is a reference component, not a production HTTP/3 implementation. The
transport binding and jurisdictional authorization model must be reviewed
before deployment.
"""

from __future__ import annotations

import base64
import json
import os
import secrets
import time
from dataclasses import dataclass
from typing import Any, Mapping

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import x25519
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF


VERSION = 1
NONCE_SIZE = 12
KEY_SIZE = 32
CAPSULE_ID_SIZE = 16


def _b64(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _unb64(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def _canonical_json(value: Mapping[str, Any]) -> bytes:
    return json.dumps(value, sort_keys=True, separators=(",", ":")).encode("utf-8")


@dataclass(frozen=True)
class CapsuleMetadata:
    capsule_id: str
    jurisdiction_id: str
    expiration: int
    emergency_endpoint_reference: str
    sender_reference: str
    authorized_contact_reference: str | None = None
    identity_registry_reference: str | None = None
    flow_context: str | None = None


class IdentityCapsule:
    """Create and open a jurisdiction-specific encrypted identity capsule."""

    @staticmethod
    def generate_recipient_keypair() -> tuple[x25519.X25519PrivateKey, x25519.X25519PublicKey]:
        private = x25519.X25519PrivateKey.generate()
        return private, private.public_key()

    @staticmethod
    def public_key_bytes(public_key: x25519.X25519PublicKey) -> bytes:
        return public_key.public_bytes(
            serialization.Encoding.Raw,
            serialization.PublicFormat.Raw,
        )

    @staticmethod
    def _derive_key(shared_secret: bytes, capsule_id: bytes) -> bytes:
        return HKDF(
            algorithm=hashes.SHA256(),
            length=KEY_SIZE,
            salt=capsule_id,
            info=b"SLeeLa/HTTP3/CIC/v1",
        ).derive(shared_secret)

    @classmethod
    def create(
        cls,
        metadata: CapsuleMetadata,
        recipient_public_key: x25519.X25519PublicKey,
        *,
        extra_attributes: Mapping[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a capsule encrypted only for the supplied recipient key.

        ``extra_attributes`` should contain only the minimum information needed
        for an authorized emergency purpose. Raw identity documents and raw
        biometric/DNA data are intentionally rejected by convention.
        """
        if metadata.expiration <= int(time.time()):
            raise ValueError("capsule expiration must be in the future")
        if not metadata.jurisdiction_id:
            raise ValueError("jurisdiction_id must not be empty")
        if not metadata.sender_reference:
            raise ValueError("sender_reference must not be empty")

        capsule_id = _unb64(metadata.capsule_id)
        if len(capsule_id) != CAPSULE_ID_SIZE:
            raise ValueError("capsule_id must encode exactly 128 bits")

        forbidden = {
            "national_id",
            "national_id_number",
            "phone_number",
            "telephone_number",
            "dna_id",
            "biometric_id",
            "raw_dna",
        }
        supplied = set((extra_attributes or {}).keys())
        forbidden_supplied = supplied & forbidden
        if forbidden_supplied:
            raise ValueError(
                "raw sensitive identifiers are not packet fields: "
                + ", ".join(sorted(forbidden_supplied))
            )

        ephemeral_private = x25519.X25519PrivateKey.generate()
        ephemeral_public = ephemeral_private.public_key()
        shared = ephemeral_private.exchange(recipient_public_key)
        key = cls._derive_key(shared, capsule_id)

        payload: dict[str, Any] = {
            "version": VERSION,
            "capsule_id": metadata.capsule_id,
            "jurisdiction_id": metadata.jurisdiction_id,
            "expiration": metadata.expiration,
            "emergency_endpoint_reference": metadata.emergency_endpoint_reference,
            "sender_reference": metadata.sender_reference,
            "authorized_contact_reference": metadata.authorized_contact_reference,
            "identity_registry_reference": metadata.identity_registry_reference,
            "flow_context": metadata.flow_context,
            "attributes": dict(extra_attributes or {}),
        }

        plaintext = _canonical_json(payload)
        nonce = os.urandom(NONCE_SIZE)
        aad = _canonical_json(
            {
                "version": VERSION,
                "capsule_id": metadata.capsule_id,
                "jurisdiction_id": metadata.jurisdiction_id,
            }
        )
        ciphertext = AESGCM(key).encrypt(nonce, plaintext, aad)

        return {
            "version": VERSION,
            "capsule_id": metadata.capsule_id,
            "jurisdiction_id": metadata.jurisdiction_id,
            "ephemeral_public_key": _b64(cls.public_key_bytes(ephemeral_public)),
            "nonce": _b64(nonce),
            "ciphertext": _b64(ciphertext),
        }

    @classmethod
    def open(
        cls,
        capsule: Mapping[str, Any],
        recipient_private_key: x25519.X25519PrivateKey,
    ) -> dict[str, Any]:
        """Decrypt a capsule with the corresponding recipient private key."""
        if int(capsule.get("version", 0)) != VERSION:
            raise ValueError("unsupported capsule version")

        capsule_id = _unb64(str(capsule["capsule_id"]))
        if len(capsule_id) != CAPSULE_ID_SIZE:
            raise ValueError("invalid capsule ID")

        ephemeral_public = x25519.X25519PublicKey.from_public_bytes(
            _unb64(str(capsule["ephemeral_public_key"]))
        )
        shared = recipient_private_key.exchange(ephemeral_public)
        key = cls._derive_key(shared, capsule_id)
        nonce = _unb64(str(capsule["nonce"]))
        aad = _canonical_json(
            {
                "version": VERSION,
                "capsule_id": capsule["capsule_id"],
                "jurisdiction_id": capsule["jurisdiction_id"],
            }
        )
        plaintext = AESGCM(key).decrypt(nonce, _unb64(str(capsule["ciphertext"])), aad)
        payload = json.loads(plaintext.decode("utf-8"))

        if payload.get("expiration", 0) <= int(time.time()):
            raise ValueError("capsule has expired")
        if payload.get("jurisdiction_id") != capsule.get("jurisdiction_id"):
            raise ValueError("jurisdiction binding failed")

        return payload


def new_capsule_id() -> str:
    """Generate a 128-bit public capsule identifier."""
    return _b64(secrets.token_bytes(CAPSULE_ID_SIZE))
